docs/guides/debugging.md

```diff
@@ -0,0 +1,226 @@
+# Debugging techniques
+
+Errors will happen, be it during builds, unit tests, integration testing, or somewhere else in your pipeline. Earthly provides an interactive debugger to aid debugging command failures.
+
+Debugging an error can be challenging during a traditional image build, since the container state disappears as soon as the build error occurs. A developer may then inject various print commands through out the build to help reason about the state of the system before the failure occurs.
+
+Let's consider a build example:
+
+```Dockerfile
+# Earthfile
+
+FROM python:3
+WORKDIR /code
+
+build:
+  RUN curl https://raw.githubusercontent.com/jsvine/markovify/master/test/texts/sherlock.txt > /sherlock.txt
+  COPY generate_phrase.py .
+  RUN pip3 install markovify
+  RUN python3 generate_phrase.py
+```
+
+and our python code:
+```Python
+# generate_phrase.py
+
+import markovify
+text = open('sherlock.txt').read()
+text_model = markovify.Text(text)
+print(text_model.make_sentence())
+```
+
+
+Now we can run it with `earth +test`, and we'll see a failure has occured:
+
+```
+=========================== FAILURE ===========================
++build *failed* | --> RUN python3 generate_phrase.py
++build *failed* | Traceback (most recent call last):
++build *failed* |   File "generate_phrase.py", line 3, in <module>
++build *failed* |     text = open('sherlock.txt').read()
++build *failed* | FileNotFoundError: [Errno 2] No such file or directory: 'sherlock.txt'
++build *failed* | Command /bin/sh -c python3 generate_phrase.py failed with exit code 1
++build *failed* | +build *failed* | ERROR: Command exited with non-zero code: RUN python3 generate_phrase.py
+Error: solve side effects: solve: failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c  /usr/bin/earth_debugger /bin/sh -c 'python3 generate_phrase.py']: buildkit-runc did not terminate successfully
+```
+
+Why can't it find the sherlock.txt file? Let's re-run `earth` with the `--interactive` (or `-i`) flag: `earth -i +test`
+
+This time we see a slighly different message:
+
+```
++build | --> RUN python3 generate_phrase.py
++build | Traceback (most recent call last):
++build |   File "generate_phrase.py", line 3, in <module>
++build |     text = open('sherlock.txt').read()
++build | FileNotFoundError: [Errno 2] No such file or directory: 'sherlock.txt'
++build | Command /bin/sh -c python3 generate_phrase.py failed with exit code 1
++build | Entering interactive debugger (**Warning: only a single debugger per host is supported**)
++build | root@buildkitsandbox:/code#
+```
+
+This time rather than exiting, earth will drop us into an interactive root shell, where we can type in arbitrary commands to figure out the problem:
```

Comment:
I would add here "into an interactive root shell, within the container of build environment" - to clarify that this shell is within the build itself.

(We used the term "build environment" in other places to describe this place)